:tractor: Parameterise rancher PR deploy script

neilbmclaughlin opened the pull request:
To allow this script to be used by different repos need to allow
aspects to be parameterised.

Added script to allow easy configuration of Travis environment
variables too.

Note: The PR script can be run locally by setting the TRAVIS env vars (e.g. `TRAVIS=yes RANCHER_TEMPLATE_NAME=c2s-pharmacy-finder TRAVIS_REPO_SLUG=nhsuk/connecting-to-services TRAVIS_PULL_REQUEST=<insert PR number here for an image which still exists> SPLUNK_HEC_TOKEN=blah GITHUB_ACCESS_TOKEN=<insert personal access token here> ./rancher_pr_deploy.sh nearby-services-api`)

neilbmclaughlin committed:
:tractor: Paramaterise rancher PR deploy script

To allow this script to be used by different repos need to allow
aspects to be parameterised.

Added script to allow easy configuration of Travis environment
variables too.

neilbmclaughlin commented on rancher_pr_deploy.sh:
```diff
@@ -17,25 +23,35 @@ if [[ -n "$TRAVIS" ]]; then
 
-    curl -s "https://raw.githubusercontent.com/nhsuk/nhsuk-rancher-templates/master/templates/${RANCHER_TEMPLATE_NAME}/0/docker-compose.yml"  -o docker-compose.yml
-    curl -s "https://raw.githubusercontent.com/nhsuk/nhsuk-rancher-templates/master/templates/${RANCHER_TEMPLATE_NAME}/0/rancher-compose.yml" -o rancher-compose.yml
+    curl -s "https://raw.githubusercontent.com/nhsuk/nhsuk-rancher-templates/feature/changes-for-ci-script/templates/${RANCHER_TEMPLATE_NAME}/0/docker-compose.yml"  -o docker-compose.yml
```
This is hard coded for the feature branch to facilitate testing and will need to be changed once the template changes are merged. There are almost certainly better ways of doing this.

neilbmclaughlin commented on rancher_pr_deploy.sh:
```diff
@@ -17,25 +23,35 @@ if [[ -n "$TRAVIS" ]]; then
+$(create_parameter_name "$REPO")_docker_image_tag=pr-${TRAVIS_PULL_REQUEST}
+splunk_hec_endpoint=https://splunk-collector.cloudapp.net:8088
+splunk_hec_token=${SPLUNK_HEC_TOKEN}
+hotjar_id=265857
```
These values are hard coded here and also in the rancher-compose.yml as defaults. Probably better, from a code duplication perspective, for them to be env vars and defined in a single place.

st3v3nhunt replied:
Yup, good point. Are you going to update these to be taken from the environment?

neilbmclaughlin replied:
Raised issue #5

neilbmclaughlin commented on configure-travis-env.sh:
```diff
@@ -0,0 +1,51 @@
+#!/bin/bash
+
+red=$'\e[1;31m'
```
I'll remove the unused ones of these when removing the changes introduced for testing.

neilbmclaughlin commented on rancher_pr_deploy.sh:
```diff
@@ -1,5 +1,11 @@
 #!/usr/bin/env bash
 
+DEPENDANT_SERVICE=$1
```
The dependent services (which is defined in the .travis.yml file) can be inferred from looking at the docker-compose.yml file but, as it stands, the differences in naming makes it non-trivial (e.g. c2s-frontend vs connecting-to-services) so this approach works for now. The variety of names we use for the same things is making it harder for us to script stuff.

st3v3nhunt replied:
Can something be added to the `README` about this? Possibly add an example into `.travis.yml` too?

neilbmclaughlin replied:
README updated.

st3v3nhunt replied:
I don't see any additional info in the `README` about `DEPENDANT_SERVICE`

neilbmclaughlin commented on rancher_pr_deploy.sh:
```diff
@@ -1,5 +1,11 @@
+DEPENDANT_SERVICE=$1
+
+create_parameter_name() {
+  echo "$1" | tr '-' '_'
```
\- is an invalid character for parameter names but not elsewhere (e.g. stack names)

neilbmclaughlin commented on rancher_pr_deploy.sh:
```diff
@@ -17,34 +23,49 @@ if [[ -n "$TRAVIS" ]]; then
+webtrends_id=dcs222rfg0jh2hpdaqwc2gmki_9r4q
+EOT
+ 
+    if [ -n "$DEPENDANT_SERVICE" ]
```
I'm expecting that when we introduce the api with mongo then, rather than just handling a single dependant service we will change this to loop through multiple ones.

st3v3nhunt commented on rancher_pr_deploy.sh:
```diff
@@ -17,34 +23,49 @@ if [[ -n "$TRAVIS" ]]; then
+ 
+    if [ -n "$DEPENDANT_SERVICE" ]
+    then
+      LATEST_RELEASE=$(curl -s "https://api.github.com/repos/nhsuk/${DEPENDANT_SERVICE}/releases/latest" | jq -r '.tag_name')
```
Is `jq` installed in the CI environment?

neilbmclaughlin replied:
Yes

st3v3nhunt commented on rancher_pr_deploy.sh:
```diff
@@ -17,34 +23,49 @@ if [[ -n "$TRAVIS" ]]; then
     PAYLOAD="{\"body\": \"${MSG}\" }"
 
-    curl -s -d "${PAYLOAD}" "https://api.github.com/repos/${TRAVIS_REPO_SLUG}/issues/${TRAVIS_PULL_REQUEST}/comments?access_token=${GITHUB_ACCESS_TOKEN}"
+    GITHUB_RESPONSE=$(curl -s -d "${PAYLOAD}" "https://api.github.com/repos/${TRAVIS_REPO_SLUG}/issues/${TRAVIS_PULL_REQUEST}/comments?access_token=${GITHUB_ACCESS_TOKEN}" | jq -r '.body')
```
Is `jq` installed in the CI environment?

neilbmclaughlin replied:
Yes.

st3v3nhunt reviewed (changes requested):
I can't see where `configure-travis-env.sh` is used. Any chance of a bit of explanation please?

Are there any linters that could be used to validate the scripts on checkin (can be raised as an issue and done separately).

neilbmclaughlin committed:
:books: Update readme

neilbmclaughlin commented:
Linting is possible but we need to work out how to install the githooks across dev machines. To be discussed in the dev catchup meeting.

Notes about `configure-travis-env.sh` added to README

neilbmclaughlin commented:
@st3v3nhunt - should be good for merging.

st3v3nhunt commented on README.md:
```diff
@@ -29,9 +29,17 @@ As the scripts are intended to be by Travis they use a number of environment var
+
+Finally, it is worth noting that it is possible to run the scripts locally by setting the envirionment variables. For example:
+
+ `TRAVIS=yes RANCHER_TEMPLATE_NAME=c2s-pharmacy-finder TRAVIS_REPO_SLUG=nhsuk/connecting-to-services TRAVIS_PULL_REQUEST=<PR number of an image which still exists> SPLUNK_HEC_TOKEN=<splunk token> GITHUB_ACCESS_TOKEN=<personal access token> ./rancher_pr_deploy.sh nearby-services-api`
```
`TRAVIS` is a boolean i.e. `true` rather than `yes` https://docs.travis-ci.com/user/environment-variables/#Default-Environment-Variables

st3v3nhunt reviewed (changes requested):
A couple of `README` related tweaks.

Regarding the versioning of the templates,  just as with the dependent services having this specified by the project running the script seems sensible. I wonder whether the non-secret environment variables should be defined in a file the running repo maintains e.g. `.travis.env`. Something like that would reduce the amount of env vars having to be added into the Travis env to just those that are secret. None of which is really about the original question of what to do about template versions. I need to read up about it but atm, I've seen new template version used for changes in the makeup of the stack e.g. an additional service. The pharmacy finder is using version `1` with the 3 service stack.

st3v3nhunt commented:
@neilbmclaughlin I'm not sure the order you are will be tackling these PRs but I think nhsuk/nhsuk-rancher-templates#14 needs to be fixed before this is finalised?
Adding WIP label

neilbmclaughlin committed:
:1234: Handle multiple dependent services

Also, default to using the image tagged 'latest' if a repo does not
have a current release.

neilbmclaughlin committed:
:bug: Add missing env var

neilbmclaughlin committed:
:shower: Tidy up code

neilbmclaughlin committed:
:bug: Fix broken default value

neilbmclaughlin committed:
:turtle: Add defensive code

neilbmclaughlin committed:
:diamonds: Extract repo names from rancher-compose.yml

The rancher-compose.yml file contains variables for each of the image
tags for the parent service and each dependent services.
Extract the dependent service names from from this file.

It does rely on the naming convention for the variables to be
maintained in that the format
i.e. {sanitisied repo name}_docker_image_tag
sanitisation takes the form of translating '-' to '_'

neilbmclaughlin commented:
I've removed the WIP tag from this and nhsuk/nhsuk-rancher-templates#14.

I think they can both be merged. I then need to update, say, connecting to services to use master in the CI submodule.

I think this order of doing things will not break the build process but I'm not certain.

st3v3nhunt commented on rancher_pr_deploy.sh:
```diff
@@ -10,41 +70,60 @@ install_rancher() {
-    curl -s -d "${PAYLOAD}" "https://api.github.com/repos/${TRAVIS_REPO_SLUG}/issues/${TRAVIS_PULL_REQUEST}/comments?access_token=${GITHUB_ACCESS_TOKEN}"
+    GITHUB_RESPONSE=$(curl -s -o /dev/null -w '%{http_code}' -d "${PAYLOAD}" "https://api.github.com/repos/${TRAVIS_REPO_SLUG}/issues/${TRAVIS_PULL_REQUEST}/comments?access_token=${GITHUB_ACCESS_TOKEN}")
+
+    if [ "${GITHUB_RESPONSE}" = "201" ]; then
```
👍

st3v3nhunt commented on rancher_pr_deploy.sh:
```diff
@@ -1,7 +1,67 @@
+}
+
+create_compose_file() {
+
```
Remove space for consistency with other functions?

st3v3nhunt commented on README.md:
```diff
@@ -37,4 +38,9 @@ As the scripts are intended to be run from within the Travis CI environment they
 
+Finally, it is worth noting that it is possible to run the scripts locally by setting the envirionment variables. For example:
+
+ `TRAVIS=true RANCHER_TEMPLATE_NAME=c2s-pharmacy-finder TRAVIS_REPO_SLUG=nhsuk/connecting-to-services TRAVIS_PULL_REQUEST=<PR number of an image which still exists> SPLUNK_HEC_TOKEN=<splunk token> GITHUB_ACCESS_TOKEN=<personal access token> ./rancher_pr_deploy.sh nearby-services-api`
```
`TEMPLATE_BRANCH_NAME` is missing from this line and also from the table of env vars above?

st3v3nhunt reviewed (changes requested):
Some small tweaks.

It would be good to have the info about the tag needing to not be marked as pre-release in the README. Possibly an extra bit to the error message that asks that question to help people understand why the latest release might not have been found?

neilbmclaughlin committed:
:book: Changes from review

* Update readme
* Fix bug with incorrectly identifying the current repo/service as a
dependent service
* Rename RANCHER_TEMPLATE_BRANCH_NAME and set a default
* Add improved error diagnostics when repo doesn't have a latest release
* Some refactoring of methods for legibility

neilbmclaughlin commented:
@st3v3nhunt review responses committed.

st3v3nhunt commented on README.md:
```diff
@@ -32,6 +35,7 @@ As the scripts are intended to be run from within the Travis CI environment they
 | `RANCHER_SECRET_KEY`   | YES     | The 'password' part of the API credentials for Rancher access [Rancher API tokens](https://docs.rancher.com/rancher/v1.1/en/api/v1/api-keys/)  | NO             |
 | `RANCHER_ENVIRONMENT`  | NO     | The environment within Rancher where the stack will be deployed to                                                                                  | NO             |
 | `RANCHER_TEMPLATE_NAME`| NO     | Name of directory where Rancher config is stored in [nhsuk-rancher-templates](https://github.com/nhsuk/nhsuk-rancher-templates) e.g. `c2s-profiles` | NO             |
+| `RANCHER_TEMPLATE_BRANCH_NAME`| NO     | For development use only. Gives the ability use a branch other than master. If not specified then will default to master. | NO             |
```
Missing 'to' in between 'ability use'

Surround `mater` with backticks?

st3v3nhunt commented on README.md:
```diff
@@ -5,10 +5,13 @@ This repo contains deployment scripts executed during project builds on [Travis
-
+* Build and tag the Docker image as detailed in the repo's `Dockerfile`. This happens on PR creation and update, Tagging (including Release) and commits to the `master` branch.
+* Push the image to the appropriate docker repo
+* If the Travis build has been triggered by a PR a new stack will be created within the dev environment in Rancher. The rancher and docker templates used to build the the stack are pulled from the [nhsuk/nhsuk-rancher-templates](https://github.com/nhsuk/nhsuk-rancher-templates) repo on GitHub. The specific compose files are determined by the env var RANCHER_TEMPLATE_NAME and the subdirectory '0'  ...
```
Double 'the' at '...build the the stack...'

Maybe surround `RANCHER_TEMPLATE_NAME` with back ticks?

st3v3nhunt commented on README.md:
```diff
@@ -41,6 +45,6 @@ As the scripts are intended to be run from within the Travis CI environment they
+Finally, it is worth noting that it is possible to run the PR script locally by setting the environment variables. For example:
 
- `TRAVIS=true RANCHER_TEMPLATE_NAME=c2s-pharmacy-finder TRAVIS_REPO_SLUG=nhsuk/connecting-to-services TRAVIS_PULL_REQUEST=<PR number of an image which still exists> SPLUNK_HEC_TOKEN=<splunk token> GITHUB_ACCESS_TOKEN=<personal access token> ./rancher_pr_deploy.sh nearby-services-api`
+ `TRAVIS=true RANCHER_TEMPLATE_NAME=c2s-pharmacy-finder TRAVIS_REPO_SLUG=nhsuk/connecting-to-services TRAVIS_PULL_REQUEST=<PR number of an image which still exists> SPLUNK_HEC_TOKEN=<splunk token> GITHUB_ACCESS_TOKEN=<personal access token> TEMPLATE_BRANCH_NAME=<usually master> ./rancher_pr_deploy.sh`
```
I think `TEMPLATE_BRANCH_NAME` should be `RANCHER_ TEMPLATE_BRANCH_NAME`?

st3v3nhunt reviewed (changes requested):
A couple of typos and possibly some highlighting (my personal preference though so totally up to you).

One change that is required is the rename of the env var in the example cmd

neilbmclaughlin committed:
:bug: Fixed some typos & changed some wording

st3v3nhunt reviewed (approved):
👍
